core/src/kmx/kmx_plus.cpp

```diff
@@ -261,8 +277,60 @@ COMP_KMXPLUS_STRS::valid(KMX_DWORD _kmn_unused(length)) const {
       assert(start[length] == 0);
       return false;
     }
-    // TODO-LDML: validate valid UTF-16LE?
     DebugLoad("strs #0x%X: '%s'", i, Debug_UnicodeString(start));
+    if (!COMP_KMXPLUS_STRS::valid_string(start, length)) {
+      DebugLog("#0x%X: String of length 0x%x invalid", i, length);
+      return false;
+    }
+  }
+  return true;
+}
+
+bool COMP_KMXPLUS_STRS::valid_string(const KMX_WCHAR* start, KMX_DWORD length) {
+  for (KMX_DWORD n = 0; n < length; n++) {
+    const auto& ch = start[n];
+    if (Uni_IsSurrogate2(ch)) {
+      DebugLog("String of length 0x%x @ 0x%x: Char U+%04X is a trailing (unpaired) surrogate", length, n, ch);
+      return false;
+    } else if (Uni_IsSurrogate1(ch)) {
+      n++;
+      if (n == length) {
+        DebugLog("String of length 0x%x @ 0x%x: Char U+%04X ends with leading (unpaired) surrogate", length, n, ch);
+        return false;
+      }
+      const auto& ch2 = start[n];
+      if (!Uni_IsSurrogate2(ch2)) {
+        DebugLog("String of length 0x%x @ 0x%x: Char U+%04X not a trailing surrogate", length, n, ch2);
+        return false;
+      }
+      const km_core_usv ch32 = Uni_SurrogateToUTF32(ch, ch2);
+      if (!Uni_IsValid(ch32)) {
+        DebugLog("String of length 0x%x @ 0x%x: Char U+%08X is illegal char", length, n, ch32);
```

Comment:
```suggestion
        DebugLog("String of length 0x%x @ 0x%x: Char U+%04X is illegal char", length, n, ch32);
```

Nit: should only zero-extend up to four digits for U+xxxx.